Initial_Jrom_Data_Understanding :
1. Dataset diversity - various postures (treading water, breaststroke, backstroke, freestyle, group configurations with 2/3/4 individuals, were included)
2. 8572 labeled images representing drowning, treading water, and swimming
3. 7000 randomly chosen for training set, 1572 allocated to validation set
4. Every image under images folder has a corresponding text file under labels folder with the same name
5. Each text file shows YOLO object detection format: class, x_center, y_center, width, height
6. However, the paper does not explicitly define what each class ID (e.g., 0, 1, 2) represents in the drowning detection task.
7. Based on a labeled image of someone drowning in the paper and observations in the dataset, we can infer that class 2 represents drowning (vertical position with both hands up).
8. Further inference from Figure 5 in the paper suggests that class 0 is swimming and class 1 is tread water

Jishen meeting notes :

Original Training Images -> ROIs -> Training

We need these steps:

1. Human Detection/Segmentation, seperation of ROIs from frames (With YOLO/YOLO-obb)
2. Human Tracking (With SORT)
3. Drowning Detection (With our own model e.g. ViT, CKAN, ResNet..... all CNN-based)
4. Compute Running Threshold (y_t = \alpha y_t + (1 - \alpha) y_(t-1))
5. Draw results

t value > 0.X --> classify as drowning

This value can also be used to compute the colour of the bounding box 

But most of these steps does not involve any classical image processing steps, so maybe we need to include something like

Histrogram Equalisation (To remove the blue background, but there are also pretrains can do this)
Posture Estimation (But not sure how it helps)

Do notice that the ground truth given is in [Prediction_Outcome, Bounding Box Coordinates]

# Train YOLO Model
Warning : Running the training code using YOLO train function will generate a lot of files in the directory

In [1]:
import os
import torch
import cv2
from tqdm.notebook import tqdm
from ultralytics import YOLO

In [2]:
# Check if CUDA gpu is available
print(f"CUDA available: {torch.cuda.is_available()}")
if torch.cuda.is_available():
    print(f"GPU: {torch.cuda.get_device_name(0)}")
    device = torch.device("cuda")
else:
    device = torch.device("cpu")
    print("Using CPU")

CUDA available: True
GPU: NVIDIA GeForce RTX 4060 Laptop GPU


In [6]:
# Project file paths
notebook_dir = os.getcwd()  # Current directory (notebooks)
project_root = os.path.dirname(notebook_dir)  # go up one level to get project root
data_dir = os.path.join(project_root, "data")
image_dir = os.path.join(data_dir, "images")
label_dir = os.path.join(data_dir, "labels")
model_dir = os.path.join(project_root, "models")

print(f"Project root: {project_root}")
print(f"Data directory: {data_dir}")
print(f"Model directory: {model_dir}")

Project root: C:\Users\jrom\DataspellProjects\Drowning-Detection
Data directory: C:\Users\jrom\DataspellProjects\Drowning-Detection\data
Model directory: C:\Users\jrom\DataspellProjects\Drowning-Detection\models


In [7]:
class_names = {0: "swimming", 1: "tread water", 2: "drowning"}

## 1. Prepare data for YOLO training
create yaml file for the dataset

In [8]:
def create_data_yaml():
    yaml_content = f"""
            # Dataset paths
            path: {data_dir}  # dataset root directory
            train: {os.path.join("images", "train")}  # train images relative to path
            val: {os.path.join("images", "val")}  # val images relative to path

            # Classes
            names:
              0: swimming
              1: tread water
              2: drowning
            """
    yaml_path = os.path.join(data_dir, "drowning_data.yaml")
    with open(yaml_path, "w") as f:
        f.write(yaml_content)
    print(f"Created data configuration at {yaml_path}")
    return yaml_path

# Create the data.yaml file
data_yaml = create_data_yaml()

Created data configuration at C:\Users\jrom\DataspellProjects\Drowning-Detection\data\drowning_data.yaml


## 2. Train YOLO11n model

In [15]:
def train_yolo_model(data_yaml, epochs=50, imgsz=640, batch_size=16):
    print("Loading YOLO11n model...")
    model = YOLO("yolo11n.pt")

    print(f"Training model for {epochs} epochs...")
    results = model.train(
        data=data_yaml,
        epochs=epochs,
        imgsz=imgsz,
        batch=batch_size,
        device=0 if torch.cuda.is_available() else 'cpu',
        project=model_dir,
        name="drowning_detection",
        exist_ok=True
    )

    # Save the trained model
    model_path = os.path.join(model_dir, "drowning_detection_yolo11n.pt")
    model.save(model_path)
    print(f"Model saved to {model_path}")

    return model, results

## 3. Validate the model

In [3]:
def validate_model(data_yaml, batch_size=4):  # Remove model argument
    print("Loading best model for validation...")
    model = YOLO(os.path.join(model_dir, "drowning_detection", "weights", "best.pt"))  # Load best model
    
    print("Validating model...")
    results = model.val(data=data_yaml, batch=batch_size, device=0 if torch.cuda.is_available() else 'cpu')

    # Print metrics
    metrics = results.box
    print("\nValidation Results:")
    print(f"mAP50: {metrics.map50.mean():.4f}")
    print(f"mAP50-95: {metrics.map.mean():.4f}")
    print(f"Precision: {metrics.p.mean():.4f}")
    print(f"Recall: {metrics.r.mean():.4f}")

    return results


## 4. Extract ROIs & save to new folder
Only if satisfied with the performance

In [17]:
def extract_and_save_rois(model, confidence=0.5):
    # Create output directory for ROIs
    roi_dir = os.path.join(data_dir, "roi_extracts")
    os.makedirs(roi_dir, exist_ok=True)

    # Create subdirectories for each class
    for class_id, class_name in class_names.items():
        os.makedirs(os.path.join(roi_dir, class_name), exist_ok=True)

    # Path to validation images
    val_img_dir = os.path.join(image_dir, "val")

    # Process each image in the validation set
    image_files = [f for f in os.listdir(val_img_dir) if f.endswith(('.jpg', '.jpeg', '.png'))]

    print(f"Extracting ROIs from {len(image_files)} images...")
    for img_file in tqdm(image_files):
        img_path = os.path.join(val_img_dir, img_file)

        # Run inference on the image
        results = model(img_path)

        # Process the results
        for result in results:
            img = cv2.imread(img_path)
            img_rgb = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)

            # Extract each detection
            boxes = result.boxes
            for i, box in enumerate(boxes):
                # Get detection details
                cls_id = int(box.cls[0].item())
                conf = box.conf[0].item()

                # Skip detections with low confidence
                if conf < confidence:
                    continue

                # Get bounding box coordinates
                x1, y1, x2, y2 = map(int, box.xyxy[0])

                # Extract ROI
                roi = img_rgb[y1:y2, x1:x2]

                # Skip if ROI is empty
                if roi.size == 0:
                    continue

                # Create filename for ROI
                base_name = os.path.splitext(img_file)[0]
                roi_filename = f"{base_name}_roi{i}_class{cls_id}_{class_names[cls_id]}.jpg"

                # Save ROI to appropriate class folder
                roi_path = os.path.join(roi_dir, class_names[cls_id], roi_filename)
                cv2.imwrite(roi_path, cv2.cvtColor(roi, cv2.COLOR_RGB2BGR))

    print(f"ROIs extracted and saved to {roi_dir}")
    return roi_dir

# Run all steps

In [18]:
# 2. Train the model
model, training_results = train_yolo_model(data_yaml, epochs=50)

Loading YOLO11n model...


100%|██████████| 5.35M/5.35M [00:00<00:00, 13.1MB/s]


Training model for 50 epochs...
New https://pypi.org/project/ultralytics/8.3.92 available  Update with 'pip install -U ultralytics'
Ultralytics 8.3.79  Python-3.12.3 torch-2.6.0+cu118 CUDA:0 (NVIDIA GeForce RTX 4060 Laptop GPU, 8188MiB)
engine\trainer: task=detect, mode=train, model=yolo11n.pt, data=C:\Users\jrom\DataspellProjects\Drowning-Detection\data\drowning_data.yaml, epochs=50, time=None, patience=100, batch=16, imgsz=640, save=True, save_period=-1, cache=False, device=0, workers=8, project=C:\Users\jrom\DataspellProjects\Drowning-Detection\models, name=drowning_detection, exist_ok=True, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source

train: Scanning C:\Users\jrom\DataspellProjects\Drowning-Detection\data\labels\train... 7000 images, 0 backgrounds, 0 corrupt: 100%|██████████| 7000/7000 [00:04<00:00, 1430.86it/s]


train: New cache created: C:\Users\jrom\DataspellProjects\Drowning-Detection\data\labels\train.cache


val: Scanning C:\Users\jrom\DataspellProjects\Drowning-Detection\data\labels\val... 1572 images, 0 backgrounds, 0 corrupt: 100%|██████████| 1572/1572 [00:02<00:00, 554.59it/s]


val: New cache created: C:\Users\jrom\DataspellProjects\Drowning-Detection\data\labels\val.cache
Plotting labels to C:\Users\jrom\DataspellProjects\Drowning-Detection\models\drowning_detection\labels.jpg... 
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.001429, momentum=0.9) with parameter groups 81 weight(decay=0.0), 88 weight(decay=0.0005), 87 bias(decay=0.0)
Image sizes 640 train, 640 val
Using 8 dataloader workers
Logging results to C:\Users\jrom\DataspellProjects\Drowning-Detection\models\drowning_detection
Starting training for 50 epochs...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       1/50      4.19G      1.449      1.964      1.152         21        640: 100%|██████████| 438/438 [01:36<00:00,  4.56it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.08it/s]


                   all       1572       2317      0.899      0.894      0.947      0.614

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       2/50      4.22G      1.322      1.095      1.101         20        640: 100%|██████████| 438/438 [01:28<00:00,  4.93it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.11it/s]

                   all       1572       2317      0.916      0.919      0.962      0.637



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       3/50      4.23G      1.297     0.9164       1.09         23        640: 100%|██████████| 438/438 [01:27<00:00,  5.03it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.06it/s]

                   all       1572       2317      0.933      0.935      0.966      0.655



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       4/50       4.2G      1.267     0.8471      1.081         10        640: 100%|██████████| 438/438 [01:26<00:00,  5.05it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.11it/s]

                   all       1572       2317      0.922      0.937      0.969      0.666



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       5/50      4.21G      1.252     0.7997      1.073         24        640: 100%|██████████| 438/438 [01:25<00:00,  5.13it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.13it/s]

                   all       1572       2317      0.929      0.943      0.964      0.675



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       6/50      4.22G      1.226     0.7704       1.06         25        640: 100%|██████████| 438/438 [01:25<00:00,  5.11it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.12it/s]

                   all       1572       2317      0.936      0.946      0.977      0.684



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       7/50      4.21G      1.207     0.7369      1.053         18        640: 100%|██████████| 438/438 [01:25<00:00,  5.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.15it/s]

                   all       1572       2317      0.957      0.959      0.983       0.69



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       8/50      4.22G      1.193     0.7196      1.048         28        640: 100%|██████████| 438/438 [01:25<00:00,  5.12it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:11<00:00,  4.19it/s]

                   all       1572       2317      0.945       0.95      0.971      0.672



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       9/50      4.22G      1.178     0.6959      1.039         24        640: 100%|██████████| 438/438 [01:26<00:00,  5.06it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.11it/s]

                   all       1572       2317      0.962       0.95      0.978        0.7



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      10/50      4.22G      1.174     0.6941      1.037         17        640: 100%|██████████| 438/438 [01:26<00:00,  5.08it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.07it/s]

                   all       1572       2317      0.938      0.967      0.972      0.704



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      11/50      4.22G      1.159      0.683      1.033         21        640: 100%|██████████| 438/438 [01:26<00:00,  5.04it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:11<00:00,  4.19it/s]

                   all       1572       2317      0.958      0.962      0.982      0.715



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      12/50      4.22G      1.158     0.6744      1.032         21        640: 100%|██████████| 438/438 [01:25<00:00,  5.12it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.11it/s]

                   all       1572       2317      0.956      0.956      0.982      0.713



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      13/50      4.21G      1.149     0.6626       1.03         17        640: 100%|██████████| 438/438 [01:25<00:00,  5.13it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.16it/s]

                   all       1572       2317      0.942      0.968      0.982      0.707



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      14/50      4.21G      1.141     0.6532       1.03         16        640: 100%|██████████| 438/438 [01:25<00:00,  5.13it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.14it/s]

                   all       1572       2317      0.948      0.954      0.978      0.717



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      15/50      4.21G      1.128     0.6428      1.021         19        640: 100%|██████████| 438/438 [01:25<00:00,  5.11it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.16it/s]

                   all       1572       2317      0.949      0.965      0.981      0.706



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      16/50      4.22G      1.121     0.6436      1.018         16        640: 100%|██████████| 438/438 [01:26<00:00,  5.04it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.12it/s]

                   all       1572       2317      0.961       0.96      0.984      0.714



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      17/50      4.21G      1.121     0.6268      1.017         17        640: 100%|██████████| 438/438 [01:25<00:00,  5.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.16it/s]

                   all       1572       2317      0.948      0.963       0.98       0.72



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      18/50      4.22G       1.12     0.6313      1.016         30        640: 100%|██████████| 438/438 [01:25<00:00,  5.12it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:11<00:00,  4.20it/s]

                   all       1572       2317      0.957      0.959      0.984       0.72



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      19/50      4.22G        1.1     0.6152      1.009         30        640: 100%|██████████| 438/438 [01:26<00:00,  5.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.11it/s]

                   all       1572       2317      0.964      0.962      0.985      0.722



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      20/50      4.22G      1.103       0.61      1.011         23        640: 100%|██████████| 438/438 [01:26<00:00,  5.04it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.09it/s]

                   all       1572       2317      0.959      0.978      0.984      0.725



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      21/50      4.21G       1.09     0.5977      1.004         20        640: 100%|██████████| 438/438 [01:26<00:00,  5.04it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.14it/s]

                   all       1572       2317      0.964      0.967      0.984      0.734



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      22/50      4.22G       1.09     0.5997      1.005         21        640: 100%|██████████| 438/438 [01:26<00:00,  5.05it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.07it/s]

                   all       1572       2317       0.96       0.97      0.985      0.737



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      23/50      4.22G       1.09     0.5947      1.007         26        640: 100%|██████████| 438/438 [01:26<00:00,  5.08it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.12it/s]

                   all       1572       2317      0.968      0.966      0.984       0.73



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      24/50      4.21G      1.086     0.5945      1.005         25        640: 100%|██████████| 438/438 [01:25<00:00,  5.11it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.14it/s]

                   all       1572       2317      0.964      0.968      0.984      0.736



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      25/50      4.23G      1.084     0.5873      1.002         25        640: 100%|██████████| 438/438 [01:25<00:00,  5.13it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.16it/s]

                   all       1572       2317      0.962       0.97      0.986      0.732



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      26/50      4.22G      1.082     0.5822     0.9997         34        640: 100%|██████████| 438/438 [01:26<00:00,  5.08it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.15it/s]

                   all       1572       2317      0.969      0.975      0.986       0.74



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      27/50      4.21G      1.068     0.5713     0.9957         21        640: 100%|██████████| 438/438 [01:26<00:00,  5.05it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:11<00:00,  4.17it/s]

                   all       1572       2317      0.962      0.964      0.985      0.732



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      28/50      4.21G      1.075     0.5715     0.9975         26        640: 100%|██████████| 438/438 [01:25<00:00,  5.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:11<00:00,  4.18it/s]

                   all       1572       2317      0.968      0.972      0.986      0.738



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      29/50      4.21G      1.066     0.5765      0.998         24        640: 100%|██████████| 438/438 [01:25<00:00,  5.12it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.10it/s]

                   all       1572       2317      0.966      0.975      0.987      0.737



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      30/50      4.22G      1.064      0.564     0.9909         21        640: 100%|██████████| 438/438 [01:26<00:00,  5.08it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.15it/s]

                   all       1572       2317       0.97      0.976      0.987      0.738



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      31/50      4.23G      1.056     0.5535     0.9907         23        640: 100%|██████████| 438/438 [01:26<00:00,  5.08it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.14it/s]

                   all       1572       2317      0.971      0.972      0.986      0.739



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      32/50      4.23G      1.055     0.5576     0.9922         29        640: 100%|██████████| 438/438 [01:25<00:00,  5.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.10it/s]

                   all       1572       2317      0.964      0.977      0.987      0.737



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      33/50      4.21G      1.045     0.5482     0.9882         28        640: 100%|██████████| 438/438 [01:26<00:00,  5.05it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.07it/s]

                   all       1572       2317      0.972      0.972      0.986      0.745



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      34/50      4.21G      1.036      0.547     0.9867         24        640: 100%|██████████| 438/438 [01:26<00:00,  5.08it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:11<00:00,  4.18it/s]

                   all       1572       2317      0.966      0.975      0.986      0.745



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      35/50      4.22G      1.041     0.5435      0.986         16        640: 100%|██████████| 438/438 [01:25<00:00,  5.11it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.10it/s]

                   all       1572       2317      0.961      0.973      0.987      0.744



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      36/50      4.21G      1.038     0.5403      0.986         24        640: 100%|██████████| 438/438 [01:25<00:00,  5.11it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.12it/s]

                   all       1572       2317      0.972      0.973      0.987      0.745



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      37/50      4.23G      1.038     0.5408     0.9824         25        640: 100%|██████████| 438/438 [01:26<00:00,  5.07it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.14it/s]

                   all       1572       2317      0.972      0.974      0.986       0.74



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      38/50      4.22G      1.027     0.5317     0.9776         22        640: 100%|██████████| 438/438 [01:26<00:00,  5.06it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.14it/s]

                   all       1572       2317      0.966      0.978      0.986      0.744



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      39/50      4.22G      1.026     0.5286     0.9791         34        640: 100%|██████████| 438/438 [01:25<00:00,  5.12it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.12it/s]

                   all       1572       2317       0.97      0.972      0.987      0.743



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      40/50      4.22G      1.026     0.5198     0.9803         28        640: 100%|██████████| 438/438 [01:26<00:00,  5.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.01it/s]

                   all       1572       2317      0.964      0.974      0.986      0.744


Closing dataloader mosaic

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      41/50      4.21G     0.9914     0.4738     0.9719         11        640: 100%|██████████| 438/438 [01:23<00:00,  5.25it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.16it/s]

                   all       1572       2317      0.968      0.974      0.986      0.744



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      42/50      4.21G     0.9748     0.4656     0.9646         10        640: 100%|██████████| 438/438 [01:22<00:00,  5.33it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.13it/s]

                   all       1572       2317      0.971      0.973      0.987      0.745



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      43/50      4.21G     0.9736     0.4596     0.9624         10        640: 100%|██████████| 438/438 [01:21<00:00,  5.36it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.12it/s]

                   all       1572       2317      0.969      0.973      0.986      0.745



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      44/50       4.2G     0.9703     0.4639     0.9612          7        640: 100%|██████████| 438/438 [01:22<00:00,  5.30it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.01it/s]

                   all       1572       2317      0.967      0.973      0.987      0.745



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      45/50      4.21G     0.9613     0.4527     0.9597          9        640: 100%|██████████| 438/438 [01:21<00:00,  5.40it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:11<00:00,  4.20it/s]

                   all       1572       2317       0.97      0.972      0.985      0.745



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      46/50      4.21G     0.9602     0.4486     0.9581         16        640: 100%|██████████| 438/438 [01:21<00:00,  5.41it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.15it/s]

                   all       1572       2317      0.974      0.975      0.986      0.746



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      47/50      4.21G     0.9571     0.4447     0.9542         13        640: 100%|██████████| 438/438 [01:22<00:00,  5.33it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.09it/s]

                   all       1572       2317      0.972      0.973      0.986      0.745



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      48/50      4.21G     0.9487     0.4403      0.955         13        640: 100%|██████████| 438/438 [01:22<00:00,  5.30it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.13it/s]

                   all       1572       2317      0.973      0.975      0.987      0.745



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      49/50      4.21G     0.9485     0.4331     0.9517         11        640: 100%|██████████| 438/438 [01:21<00:00,  5.34it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.13it/s]

                   all       1572       2317      0.975      0.971      0.986      0.745



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      50/50      4.21G      0.943     0.4338     0.9529         11        640: 100%|██████████| 438/438 [01:22<00:00,  5.29it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:12<00:00,  4.15it/s]

                   all       1572       2317      0.974      0.974      0.987      0.748



50 epochs completed in 1.380 hours.
Optimizer stripped from C:\Users\jrom\DataspellProjects\Drowning-Detection\models\drowning_detection\weights\last.pt, 5.4MB
Optimizer stripped from C:\Users\jrom\DataspellProjects\Drowning-Detection\models\drowning_detection\weights\best.pt, 5.4MB

Validating C:\Users\jrom\DataspellProjects\Drowning-Detection\models\drowning_detection\weights\best.pt...
Ultralytics 8.3.79  Python-3.12.3 torch-2.6.0+cu118 CUDA:0 (NVIDIA GeForce RTX 4060 Laptop GPU, 8188MiB)
YOLO11n summary (fused): 100 layers, 2,582,737 parameters, 0 gradients, 6.3 GFLOPs


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 50/50 [00:13<00:00,  3.83it/s]


                   all       1572       2317      0.974      0.974      0.987      0.749
              swimming        991       1381      0.987      0.993      0.994      0.832
           tread water        502        571      0.965      0.974      0.986      0.761
              drowning        353        365      0.968      0.956      0.979      0.653
Speed: 0.2ms preprocess, 1.7ms inference, 0.0ms loss, 1.6ms postprocess per image
Results saved to C:\Users\jrom\DataspellProjects\Drowning-Detection\models\drowning_detection
Model saved to C:\Users\jrom\DataspellProjects\Drowning-Detection\models\drowning_detection_yolo11n.pt


In [9]:
# 3. Validate the model
validation_results = validate_model(data_yaml)

Loading best model for validation...
Validating model...
Ultralytics 8.3.79  Python-3.12.3 torch-2.6.0+cu118 CUDA:0 (NVIDIA GeForce RTX 4060 Laptop GPU, 8188MiB)
YOLO11n summary (fused): 100 layers, 2,582,737 parameters, 0 gradients, 6.3 GFLOPs


val: Scanning C:\Users\jrom\DataspellProjects\Drowning-Detection\data\labels\val.cache... 1572 images, 0 backgrounds, 0 corrupt: 100%|██████████| 1572/1572 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 393/393 [00:16<00:00, 23.56it/s]


                   all       1572       2317      0.974      0.974      0.987      0.749
              swimming        991       1381      0.987      0.993      0.994      0.831
           tread water        502        571      0.965      0.974      0.986      0.762
              drowning        353        365      0.969      0.956      0.979      0.653
Speed: 0.3ms preprocess, 4.3ms inference, 0.0ms loss, 1.3ms postprocess per image
Results saved to runs\detect\val

Validation Results:
mAP50: 0.9867
mAP50-95: 0.7486
Precision: 0.9738
Recall: 0.9745


high accuracy (mAP50: 98.67%), moderate generalisation across IoU thresholds (mAP50-95: 74.86%), strong precision (97.38%), and high recall (97.45%). It detects drowning instances effectively with minimal false positives and false negatives. 

In [ ]:
# 4. Extract ROIs
roi_dir = extract_and_save_rois(model, confidence=0.5)